File: 0_3.leetcode/python/LEET617.py
from typing import *
import collections

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def mergeTrees(root1: TreeNode, root2: TreeNode) -> TreeNode:
    if root1 and root2:
        node = TreeNode(root1.val +  root2.val)

        node.left = mergeTrees(root1.left, root2.left)
        node.right = mergeTrees(root1.right, root2.right)

        return node
    else:
        return root1 or root2

def mergeTrees2(root1, root2):
    if not(root1 and root2):
        return root1 or root2

    root = TreeNode(root1.val + root2.val)
    q = collections.deque([[root, root1, root2]])
    while q:
        r, t1, t2 = q.popleft()

        if t1.left and t2.left:
            node = TreeNode(t1.left.val + t2.left.val)
            q.append([node, t1.left, t2.left])
        else:
            node = t1.left or t2.left
        r.left = node

        
        if t1.right and t2.right:
            node = TreeNode(t1.right.val + t2.right.val)
            q.append([node, t1.right, t2.right])
        else:
            node = t1.right or t2.right
        r.right = node
    
    return root
def solve():
    root1 = build_tree([1,3,2,5])
    root2 = build_tree([2,1,3,None,4,None,7])

    res = mergeTrees(root1, root2)
    print_tree(res)

def build_tree(root):
    nodes = [TreeNode(r) if r is not None else None for r in root]
    
    for i in range(len(nodes)):
        node = nodes[i]
        left, right = i * 2 + 1, i * 2 + 2 

        if left < len(nodes):
            if nodes[left] is None: continue
            node.left = nodes[left]
        if right < len(nodes):
            if nodes[right] is None: continue
            node.right = nodes[right]
    return nodes[0]

def print_tree(root):
    q = collections.deque()
    q.append(root)
    res = []
    while q:
        for _ in range(len(q)):
            cur_root = q.popleft()
            res.append(cur_root.val)

            if cur_root.left:
                q.append(cur_root.left)
            if cur_root.right:
                q.append(cur_root.right)
    
    print(*res)

solve()

File: 0_3.leetcode/java/LEET617.java
/**
 * LEET617
 */
import java.util.*;
public class LEET617 {    
//   Definition for a binary tree node.
    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    public static void main(String[] args){
        TreeNode root1 = buildTree(new Integer[] {1,3,2,5});
        TreeNode root2 = buildTree(new Integer[] {2,1,3,null,4,null,7});
        System.out.println(mergeTrees(root1, root2));
    }
    public static TreeNode buildTree(Integer[] root){
        TreeNode[] nodes = new TreeNode[root.length];
        for(int i=0;i<root.length;i++){
            if(root[i] == null){
                nodes[i] = null;
            } else {
                nodes[i] = new TreeNode(root[i]);
            }
        }

        for(int i=0;i<nodes.length;i++){
            if(nodes[i] == null) continue;
            TreeNode cur_node = nodes[i];
            int left = i*2+1, right = i*2+2;

            if(left < nodes.length && nodes[left] != null){
                cur_node.left = nodes[left];
            }
            if(right < nodes.length && nodes[right] != null){
                cur_node.right = nodes[right];
            }
        }
        return nodes[0];
    }
    public static TreeNode mergeTrees(TreeNode root1, TreeNode root2){
        return dfs(root1, root2);
    }
    public static TreeNode dfs(TreeNode t1, TreeNode t2){
        if(t1 != null && t2 != null){
            TreeNode node = new TreeNode(t1.val + t2.val);
            
            node.left = dfs(t1.left, t2.left);
            node.right = dfs(t1.right, t2.right);

            return node;
        } else {
            if (t1 != null)
                return t1;
            return t2;
        }
    }
    public static TreeNode mergeTrees2(TreeNode root1, TreeNode root2){
        if(root1 == null)
            return root2;
        if(root2 == null)
            return root1;
        Deque<TreeNode> q = new ArrayDeque<>();
        TreeNode root = new TreeNode(root1.val + root2.val);
        q.add(root);
        q.add(root1);
        q.add(root2);

        while(!q.isEmpty()){
            TreeNode r = q.removeFirst();
            TreeNode t1 = q.removeFirst();
            TreeNode t2 = q.removeFirst();

            if(t1.left != null && t2.left != null){
                r.left = new TreeNode(t1.left.val + t2.left.val);
                q.add(r.left); q.add(t1.left); q.add(t2.left);
            } else if(t1.left != null){
                r.left = t1.left;
            } else {
                r.left = t2.left;
            }

            if(t1.right != null && t2.right != null){
                r.right = new TreeNode(t1.right.val + t2.right.val);
                q.add(r.right); q.add(t1.right); q.add(t2.right);
            } else if(t1.right != null){
                r.right = t1.right;
            } else {
                r.right = t2.right;
            }
        }
        return root;
    }
}